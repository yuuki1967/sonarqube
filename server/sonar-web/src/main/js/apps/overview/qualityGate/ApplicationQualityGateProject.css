/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.application-quality-gate-project {
  padding: 10px;
}

.overview-quality-gate-condition:hover .application-quality-gate-project {
  padding: 9px;
  border-width: 1px;
  border-style: solid;
  border-color: inherit;
}

.application-quality-gate-project-conditions {
  margin-top: 4px;
}

.application-quality-gate-project-conditions > li {
  display: flex;
  justify-content: space-between;
  padding-top: 2px;
  padding-bottom: 2px;
}

.application-quality-gate-project-conditions > li.is-on-leak {
  margin-left: -9px;
  margin-right: -9px;
  padding-left: 9px;
  padding-right: 9px;
  background-color: var(--leakColor);
}
