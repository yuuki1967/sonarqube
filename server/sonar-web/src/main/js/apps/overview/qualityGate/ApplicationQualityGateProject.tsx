/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import { Link } from 'react-router';
import { getLocalizedMetricName, translate } from '../../../helpers/l10n';
import { formatMeasure, isDiffMetric } from '../../../helpers/measures';
import { getProjectUrl } from '../../../helpers/urls';
import './ApplicationQualityGateProject.css';
import { ApplicationProject, ConditionAnalysis } from '../../../api/quality-gates';

interface Props {
  metrics: { [key: string]: T.Metric };
  project: ApplicationProject;
}

export default class ApplicationQualityGateProject extends React.PureComponent<Props> {
  renderCondition = (condition: ConditionAnalysis) => {
    const metric = this.props.metrics[condition.metric];
    const metricName = getLocalizedMetricName(metric);
    const threshold = condition.errorThreshold || condition.warningThreshold;
    const isDiff = isDiffMetric(condition.metric);

    return (
      <li className={classNames({ 'is-on-leak': isDiff })} key={condition.metric}>
        <span className="text-limited">
          <strong>{formatMeasure(condition.value, metric.type)}</strong> {metricName}
          {!isDiff && condition.onLeak && ' ' + translate('quality_gates.conditions.new_code')}
        </span>
        <span
          className={classNames('big-spacer-left', {
            'text-danger': condition.status === 'ERROR',
            'text-warning': condition.status === 'WARN'
          })}>
          {metric.type === 'RATING'
            ? translate('quality_gates.operator', condition.comparator, 'rating')
            : translate('quality_gates.operator', condition.comparator)}{' '}
          {formatMeasure(threshold, metric.type)}
        </span>
      </li>
    );
  };

  render() {
    const { project } = this.props;

    return (
      <Link
        className={classNames(
          'overview-quality-gate-condition',
          'overview-quality-gate-condition-' + project.status.toLowerCase()
        )}
        to={getProjectUrl(project.key)}>
        <div className="application-quality-gate-project">
          <h4>{project.name}</h4>
          <ul className="application-quality-gate-project-conditions">
            {project.conditions.filter(c => c.status !== 'OK').map(this.renderCondition)}
          </ul>
        </div>
      </Link>
    );
  }
}
