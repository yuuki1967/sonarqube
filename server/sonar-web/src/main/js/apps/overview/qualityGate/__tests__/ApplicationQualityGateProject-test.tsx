/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import ApplicationQualityGateProject from '../ApplicationQualityGateProject';

const metrics = {
  bugs: { id: '1', key: 'bugs', name: 'Bugs', type: 'INT' },
  new_coverage: { id: '2', key: 'new_coverage', name: 'Coverage on New Code', type: 'PERCENT' },
  skipped_tests: { id: '3', key: 'skipped_tests', name: 'Skipped Tests', type: 'INT' }
};

it('renders', () => {
  const project = {
    key: 'foo',
    name: 'Foo',
    status: 'ERROR',
    conditions: [
      {
        status: 'ERROR',
        metric: 'new_coverage',
        comparator: 'LT',
        onLeak: true,
        errorThreshold: '85',
        value: '82.50562381034781'
      },
      {
        status: 'WARN',
        metric: 'bugs',
        comparator: 'GT',
        onLeak: false,
        warningThreshold: '0',
        value: '17'
      },
      {
        status: 'ERROR',
        metric: 'bugs',
        comparator: 'GT',
        onLeak: true,
        warningThreshold: '0',
        value: '3'
      },
      {
        status: 'OK',
        metric: 'skipped_tests',
        comparator: 'GT',
        onLeak: false,
        warningThreshold: '0',
        value: '0'
      }
    ]
  };
  const wrapper = shallow(<ApplicationQualityGateProject metrics={metrics} project={project} />);
  expect(wrapper).toMatchSnapshot();
});
