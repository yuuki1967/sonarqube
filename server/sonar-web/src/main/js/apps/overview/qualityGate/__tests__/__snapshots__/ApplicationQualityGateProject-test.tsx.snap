// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Link
  className="overview-quality-gate-condition overview-quality-gate-condition-error"
  onlyActiveOnIndex={false}
  style={Object {}}
  to={
    Object {
      "pathname": "/dashboard",
      "query": Object {
        "branch": undefined,
        "id": "foo",
      },
    }
  }
>
  <div
    className="application-quality-gate-project"
  >
    <h4>
      Foo
    </h4>
    <ul
      className="application-quality-gate-project-conditions"
    >
      <li
        className="is-on-leak"
        key="new_coverage"
      >
        <span
          className="text-limited"
        >
          <strong>
            82.5%
          </strong>
           
          Coverage on New Code
        </span>
        <span
          className="big-spacer-left text-danger"
        >
          quality_gates.operator.LT
           
          85.0%
        </span>
      </li>
      <li
        className=""
        key="bugs"
      >
        <span
          className="text-limited"
        >
          <strong>
            17
          </strong>
           
          Bugs
        </span>
        <span
          className="big-spacer-left text-warning"
        >
          quality_gates.operator.GT
           
          0
        </span>
      </li>
      <li
        className=""
        key="bugs"
      >
        <span
          className="text-limited"
        >
          <strong>
            3
          </strong>
           
          Bugs
           quality_gates.conditions.new_code
        </span>
        <span
          className="big-spacer-left text-danger"
        >
          quality_gates.operator.GT
           
          0
        </span>
      </li>
    </ul>
  </div>
</Link>
`;
